File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int firstOccurrence = this.findLowerBound(nums, target);

        if (firstOccurrence == -1) {
            return new int[] { -1, -1 };
        }

        int lastOccurrence = this.findUpperBound(nums, target);

        return new int[] { firstOccurrence, lastOccurrence };
    }

    private int findLowerBound(int[] nums, int target) {
        int N = nums.length;
        int begin = 0, end = N - 1;

        while (begin <= end) {
            int mid = (begin + end) / 2;

            if (nums[mid] == target) {
                // This means we found our lower bound.
                if (mid == begin || nums[mid - 1] != target) {
                    return mid;
                }

                // Search on the left side for the bound.
                end = mid - 1;
            } else if (nums[mid] > target) {
                end = mid - 1;
            } else {
                begin = mid + 1;
            }
        }

        return -1;
    }

    private int findUpperBound(int[] nums, int target) {
        int N = nums.length;
        int begin = 0, end = N - 1;

        while (begin <= end) {
            int mid = (begin + end) / 2;

            if (nums[mid] == target) {
                // This means we found our upper bound.
                if (mid == end || nums[mid + 1] != target) {
                    return mid;
                }

                // Search on the right side for the bound.
                begin = mid + 1;
            } else if (nums[mid] > target) {
                end = mid - 1;
            } else {
                begin = mid + 1;
            }
        }

        return -1;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        first=self.findFirstOccurance(nums, target)
        if first==-1:
            return [-1, -1]
        last=self.findLastOccurance(nums, target)

        return [first, last]
    
    def findFirstOccurance(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        ans=-1
        while left <= right:
            mid = (left + right) // 2
            
            if nums[mid]==target:
                ans=mid
                right=mid-1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid -1

        return ans  

    def findLastOccurance(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        ans=-1
        while left <= right:
            mid = (left + right) // 2
            
            if nums[mid]==target:
                ans=mid
                left=mid+1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid -1

        return ans  